cryptography/src/transcript.rs

```diff
@@ -0,0 +1,107 @@
+use rand_core::{CryptoRngCore, OsRng};
+
+/// This struct provides a convenient abstraction over hashing data and deriving randomness.
+///
+/// It automatically takes care of details like:
+/// - correctly segmenting packets of data,
+/// - domain separating different uses of tags and randomness,
+/// - making sure that secret state is zeroized as necessary.
+struct Transcript {}
+
+impl Transcript {
+    /// Create a new transcript.
+    ///
+    /// The namespace serves to disamiguate two transcripts, so that even if they record
+    /// the same information, the results will be different:
+    /// ```
+    /// let s1 = Transcript::new(b"n1").record(b"A").summarize();
+    /// let s2 = Transcript::new(b"n2").record(b"A").summarize();
+    /// assert_neq!(s1, s2);
+    /// ```
+    pub fn new(namespace: &[u8]) -> Self {
+        todo!()
+    }
+
+    /// Start a transcript from a summary.
+    ///
+    /// Note that this will not produce the same result as if the transcript
+    /// were never summarized to begin with.
```

Comment:
Hard for me to understand the point if this is not the case ?

Reply:
I think of this method as a more directed and fool proof way of taking a `Summary` and adding it to a transcript. It also means that you don't need a namespace, since you know that one was provided to create the initial summary.

This is certainly a bit less useful than resumption being exactly the same, but this is a property that you can't easily provide. (This is because the partial state of a Hasher might be much larger than a summary, creating the summary basically erases this partial state, which is also a desirable property, since now any sensitive information that you might have been hashing is now out of the partial state, and has gone through a one way summarization function.)